contracts/gateway/AxelarAmplifierGateway.sol

```diff
@@ -92,54 +88,65 @@ contract AxelarAmplifierGateway is IAxelarAmplifierGateway {
-        bytes32 batchHash = ECDSA.toEthSignedMessageHash(keccak256(data));
-
-        bool isLatestSigners = authModule.validateProof(batchHash, signedBatch.proof);
+    function verifyMessages(SignedMessageBatch calldata signedBatch) public {
+        bytes memory data = abi.encode(signedBatch.messages);
 
-        if (domainSeparator != signedBatch.batch.domainSeparator) revert InvalidDomainSeparator();
+        _verifyProof(CommandType.VerifyMessages, data, signedBatch.proof);
 
-        Command[] calldata commands = batch.commands;
+        Message[] calldata messages = signedBatch.messages;
 
-        for (uint256 i; i < commands.length; ++i) {
-            Command calldata command = commands[i];
-            bytes32 commandId = keccak256(bytes(command.messageId));
+        for (uint256 i; i < messages.length; ++i) {
+            Message calldata message = messages[i];
+            bytes32 commandId = keccak256(bytes(message.messageId));
 
             // Ignore if commandId is already executed
             if (isCommandExecuted(commandId)) {
                 continue;
             }
 
-            if (command.commandType == CommandType.ApproveContractCall) {
-                _approveContractCall(commandId, command);
-            } else if (command.commandType == CommandType.RotateSigners) {
-                if (!isLatestSigners) {
-                    continue;
-                }
-
-                isLatestSigners = false;
-
-                _rotateSigners(command.params);
-            } else {
-                revert InvalidCommand();
-            }
+            _approveMessage(commandId, message);
 
             _storage().commands[commandId] = true;
 
             // slither-disable-next-line reentrancy-events
-            emit Executed(commandId, command.messageId);
+            emit Executed(commandId);
         }
     }
 
+    function rotateSigners(SignedRotation calldata signedRotation) public {
+        bytes memory data = abi.encode(signedRotation.rotation);
+        // TODO: prefix with CommandType for domain separation
+        bytes32 commandId = keccak256(data);
+
+        if (isCommandExecuted(commandId)) {
+            revert CommandAlreadyExecuted(commandId);
+        }
+
+        bool isLatestSigners = _verifyProof(CommandType.RotateSigners, data, signedRotation.proof);
+        if (!isLatestSigners) {
+            revert NotLatestSigners();
+        }
+
+        authModule.rotateSigners(signedRotation.rotation.newSigners);
+
+        emit OperatorshipTransferred(data);
+    }
+
     /**********************\
     |* Internal Functions *|
     \**********************/
 
+    function _verifyProof(CommandType commandType, bytes memory data, Proof calldata proof) internal view returns (bool) {
+        if (domainSeparator != proof.domainSeparator) revert InvalidDomainSeparator();
+
+        // TODO: use SignData struct
+        data = abi.encode(commandType, proof.domainSeparator, proof.signerCommitment, data);
```

Comment:
Seems odd that we have to double encode data? Could we move to something like?
```
abi.encode(commandType, proof.domainSeparator, proof.signerCommitment, signedRotation.rotation);
```
Also would be nice to have it encoded in the same sequence in the calldata so we could slice it out.

Reply:
Lets talk and think more about those structs

Reply:
yeah, haven't settled on that. the encoding needs to be done in amplifier as well. It might be easier to have the last arg as `bytes` to reuse encoding between different command types